paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_identity_inplace_op_pass.cc

```diff
@@ -0,0 +1,354 @@
+        if (is_read_by_other_ops.count(cur_var) > 0) {
+          var_is_read_by_other_ops = true;
+        }
+
+        if (iter->second.size() > 1) {
+          var_is_branched_reuse = true;
+        }
+
+        for (auto &ovar_op : iter->second) {
+          q.push(ovar_op.first);
+        }
+      }
+    }
+
+    if (var_is_read_by_other_ops || var_is_branched_reuse) {
+      unreusable_vars.insert(leaves.begin(), leaves.end());
+    }
+  }
+  return unreusable_vars;
+}
+
+class BufferSharedIdentityInplaceOpPass : public MemoryReusePass {
+ protected:
+  std::string ReuseType() const override { return "identity_inplace"; }
+
+  void Run(Graph *graph) const override;
+
+ private:
+  std::pair<IdentityInplaceVarInfo, std::unordered_set<VarHandle *>>
+  RunOnEachScope(const std::unordered_map<std::string, LastLiveOpOfVarInfo>
+                     &var_infos) const;
+
+  static VarHandle *TryFindMatchedIdentityInplacedOutput(
+      const VarHandle &in_var, const ComputationOpHandle &op);
+};
+
+std::pair<IdentityInplaceVarInfo, std::unordered_set<VarHandle *>>
+BufferSharedIdentityInplaceOpPass::RunOnEachScope(
+    const std::unordered_map<std::string, LastLiveOpOfVarInfo> &var_infos)
+    const {
+  IdentityInplaceVarInfo identity_inplace_info;
+  std::unordered_set<VarHandle *> is_read_by_other_ops;
+  std::unordered_set<VarHandle *> collected_out_vars;
+
+  /**
+   * The memory of some out var has been reused by non-identity ops,
+   * this vars must be forced to be leaf when identity inplaced is performed.
+   * If branched identity inplace is performed (many vars may reuse the memory
+   * of the same var), this kind of leaf nodes must be pruned (otherwise,
+   * caculation result may be wrong).
+   * If non-branched identity inplace is performed, this kind of leaf nodes
+   * should not be pruned.
+   */
+  std::unordered_set<VarHandle *> reused_out_vars;
+  for (auto &pair : var_infos) {
+    auto *in_var = pair.second.var();
+    if (!IsInVarReusable(*in_var)) continue;
+
+    // force leaf vars
+    if (reused_out_vars.count(in_var) > 0) continue;
+
+    auto in_arg = in_var->Name();
+
+    auto last_live_ops = pair.second.ops();
+    /**
+     * If all `ops` only read `in_var` and do not write `in_var`,
+     * we should scan all output nodes of `in_var`. If any output
+     * node of `in_var` is an identity op node, in-place can be
+     * performed.
+     */
+    bool failure = false;
+    for (auto *op : last_live_ops) {
+      failure = FindNodesByName(in_arg, op->Node()->inputs).empty() ||
+                !FindNodesByName(in_arg, op->Node()->outputs).empty();
+      if (failure) break;
+    }
+
+    if (failure) continue;
+
+    std::unordered_set<ComputationOpHandle *> candidate_ops;
+    for (auto *op_node : in_var->Node()->outputs) {
+      auto *op = dynamic_cast<ComputationOpHandle *>(
+          &(op_node->Wrapper<OpHandleBase>()));
+      if (op == nullptr) {
+        candidate_ops.clear();
+        break;
```

Comment:
Why use break here?

Reply:
Now we require that all last live ops that read `in_var` must be `ComputationOpHandle` but not something like `AllReduceOpHandle`. This is something like what `ReferenceCountPass` does.

Reply:
Got it, thanks.